test/src/native_updater_test.dart

```diff
@@ -0,0 +1,157 @@
+import 'package:flutter/cupertino.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter/services.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:native_updater/src/native_updater.dart';
+import 'package:in_app_update/in_app_update.dart';
+import 'package:native_updater/src/update_cupertino_alert.dart';
+
+void main() {
+  TestWidgetsFlutterBinding.ensureInitialized();
+  NativeUpdater nativeUpdater;
+  MethodChannel inAppUpdateChannel = const MethodChannel("in_app_update");
+  bool isImmediateUpdateInvoked;
+  bool isFlexibleUpdateInvoked;
+
+  setUp(() {
+    nativeUpdater = NativeUpdater();
+    isImmediateUpdateInvoked = false;
+    isFlexibleUpdateInvoked = false;
+
+    inAppUpdateChannel.setMockMethodCallHandler((MethodCall methodCall) async {
+      if (methodCall.method == 'checkForUpdate') {
+        return {
+          'updateAvailable': true,
+          'immediateAllowed': true,
+          'flexibleAllowed': true,
+          'availableVersionCode': 10,
+        };
+      }
+
+      if (methodCall.method == 'performImmediateUpdate') {
+        isImmediateUpdateInvoked = true;
+      }
+
+      if (methodCall.method == 'startFlexibleUpdate') {
+        isFlexibleUpdateInvoked = true;
+      }
+
+      return null;
+    });
+  });
+
+  group('Native Updater Test', () {
+    bool tForceUpdate;
+
+    group('Android', () {
```

Comment:
tests for Android are incomplete.

Tests that I would do:
Group 1, while on API level  < 21:
- If force update
   - App pops up a material alert with 2 options (got to store and close app)
   - App exits after ignoring the update
- If not force update
  - App pops up a material alert with 2 options (dismiss and go to store)
  - Dialogue is dismissable

Group 2, while on  API level  >= 21: (Here we delegate and trust the internal API, so fewer tests are needed)
- If force update
   - call internal API with the force update method
- If not force update
   - Call internal API without force update 

To achieve SDK level test you need to do something similar to 
http://robolectric.org/configuring/ see Configure SDK Level

Reply:
Will be implemented next.

Reply:
I think that we dont need to do the integration test and make it very complicated. We at least need to have a test that fakes the API level, similar to:
https://github.com/flutter/flutter/blob/1376746237e968eadae89927001fe02ac5299bf0/packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart#L133

Something that forces our source code to contain an IF like this:
https://github.com/flutter/plugins/blob/91e1c385a465c2b93c01bee274b955116b4eee9f/packages/camera/android/src/main/java/io/flutter/plugins/camera/CameraPlugin.java#L102
or this:
https://github.com/flutter/plugins/blob/91e1c385a465c2b93c01bee274b955116b4eee9f/packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/FileUtils.java#L54

Unfortunately, the plugins tests are poor :(

Reply:
Maybe it's not possible to do inside the code itself, but we could test it with gradle building the correct environment for us like in here:
https://pub.dev/packages/e2e#android-device-testing